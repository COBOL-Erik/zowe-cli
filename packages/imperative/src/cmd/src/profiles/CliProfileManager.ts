/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { BasicProfileManager } from "../../../profiles";
import { ImperativeError } from "../../../error";
import { ICommandProfileTypeConfiguration } from "../doc/profiles/definition/ICommandProfileTypeConfiguration";
import { IProfileLoaded } from "../../../profiles/src/doc";
import { ICliLoadAllProfiles } from "../doc/profiles/parms/ICliLoadAllProfiles";

/**
 * A profile management API compatible with transforming command line arguments into
 * profiles
 * @internal
 */
export class CliProfileManager extends BasicProfileManager<ICommandProfileTypeConfiguration> {

    /**
     * NOTE: This is just a copy of BasicProfileManager.loadAll
     * REASON: We needed the Abstract profile manager to call the CLI profile manager to handle loading of secure properties
     * Loads all profiles from every type. Profile types are determined by reading all directories within the
     * profile root directory.
     * @returns {Promise<IProfileLoaded[]>} - The list of all profiles for every type
     */
    public async loadAll(params?: ICliLoadAllProfiles): Promise<IProfileLoaded[]> {
        this.log.trace(`Loading all profiles for type "${this.profileType}"...`);
        // Load all the other profiles for other types
        const loadAllProfiles: any[] = [];

        // Load only the profiles for the type if requested
        if (params != null && params.typeOnly) {
            const names: string[] = this.getAllProfileNames();
            for (const name of names) {
                loadAllProfiles.push(this.load({
                    name,
                    failNotFound: true,
                    loadDependencies: false,
                    noSecure: params.noSecure
                }));
            }
        } else {

            // Otherwise, load all profiles of all types
            for (const typeConfig of this.profileTypeConfigurations) {
                const typeProfileManager = new CliProfileManager({
                    profileRootDirectory: this.profileRootDirectory,
                    typeConfigurations: this.profileTypeConfigurations,
                    type: typeConfig.type,
                    logger: this.log,
                    loadCounter: this.loadCounter
                });

                // Get all the profile names for the type and attempt to load every one
                const names: string[] = typeProfileManager.getAllProfileNames();
                for (const name of names) {
                    this.log.debug(`Loading profile "${name}" of type "${typeConfig.type}".`);
                    loadAllProfiles.push(typeProfileManager.load({
                        name,
                        failNotFound: true,
                        loadDependencies: false,
                        noSecure: (params != null) ? params.noSecure : undefined
                    }));
                }
            }
        }

        // Construct the full list for return
        let allProfiles: IProfileLoaded[] = [];
        try {
            this.log.trace(`Awaiting all loads...`);
            const theirProfiles = await Promise.all(loadAllProfiles);
            for (const theirs of theirProfiles) {
                allProfiles = allProfiles.concat(theirs);
            }
            this.log.trace(`All loads complete.`);
        } catch (e) {
            this.log.error(e.message);
            throw new ImperativeError({msg: e.message, additionalDetails: e.additionalDetails, causeErrors: e});
        }

        return allProfiles;
    }
}
