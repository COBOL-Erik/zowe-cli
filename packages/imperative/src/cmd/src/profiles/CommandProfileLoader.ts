/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { Arguments } from "yargs";
import { ICommandDefinition } from "../doc/ICommandDefinition";
import { IProfile, IProfileLoaded } from "../../../profiles";
import { CommandProfiles } from "./CommandProfiles";
import { inspect } from "util";
import { ICommandProfileLoaderParms } from "../doc/profiles/parms/ICommandProfileLoaderParms";
import { Logger } from "../../../logger";
import { ImperativeExpect } from "../../../expect";

/**
 * The command processor profile loader loads all profiles that are required (or optional) given a command
 * definitions requirements. It returns the CommandProfiles object (which contains the map and getters for the
 * command handlers usage).
 * @internal
 * @class CommandProfileLoader
 * @deprecated Use the `V1ProfileRead` class if you still need to read V1 profiles
 */
export class CommandProfileLoader {
    /**
     * Create a new instance of the profile loader
     * @static
     * @param {ICommandProfileLoaderParms} parms - contains command definition and logger
     * @returns
     * @memberof CommandProfileLoader
     */
    public static loader(parms: ICommandProfileLoaderParms) {
        // eslint-disable-next-line deprecation/deprecation
        return new CommandProfileLoader(parms.commandDefinition, parms.logger || Logger.getImperativeLogger());
    }

    /**
     * The input command definition for the command being issued.
     * @private
     * @type {ICommandDefinition}
     * @memberof CommandProfileLoader
     */
    private mCommandDefinition: ICommandDefinition;

    /**
     * Logger - supplied on the constructor - but defaults to the Imperative logger.
     * @private
     * @type {Logger}
     * @memberof CommandProfileLoader
     */
    private mLog: Logger;

    /**
     * Creates an instance of CommandProfileLoader.
     * @param {ICommandDefinition} commandDefinition - The input command definition for the command being issued.
     * @param {any} [logger=Logger.getImperativeLogger()] - A log4js instance
     * @memberof CommandProfileLoader
     */
    constructor(commandDefinition: ICommandDefinition, logger = Logger.getImperativeLogger()) {
        const err: string = "Could not construct the profile loader.";
        ImperativeExpect.toNotBeNullOrUndefined(commandDefinition, `${err} No command definition supplied.`);
        this.mCommandDefinition = commandDefinition;
        ImperativeExpect.toBeEqual(logger instanceof Logger, true, `${err} The "logger" supplied is not of type Logger.`);
        this.mLog = logger;
        this.log.trace(`Profile loader created for command: ${commandDefinition.name}`);
    }

    /**
     * Load the profiles for the command - the command arguments are supplied to grab the profile names from
     * the arguments supplied by the user.
     * @param {Arguments} commandArguments - The command arguments supplied on this command invocation (Yargs style)
     * @returns {Promise<CommandProfiles>} - The promise is fulfilled with the map object OR rejected with an
     * Imperative error
     * @memberof CommandProfileLoader
     */
    // eslint-disable-next-line deprecation/deprecation
    public async loadProfiles(commandArguments: Arguments): Promise<CommandProfiles> {
        // Validate parms
        ImperativeExpect.toNotBeNullOrUndefined(commandArguments, `Could not load profiles. No command arguments supplied.`);

        // Log the API call
        this.log.info(`Request to load profiles for command: ${this.definition.name}...`);
        this.log.trace(`Profile load arguments supplied:\n${inspect(commandArguments, {depth: null})}`);

        // Create the map that eventually will be returned
        const profileMap: Map<string, IProfile[]> = new Map<string, IProfile[]>();
        const profileMetaMap: Map<string, IProfileLoaded[]> = new Map<string, IProfileLoaded[]>();

        // We no longer read V1 profile files, so just return empty maps
        // eslint-disable-next-line deprecation/deprecation
        return new CommandProfiles(profileMap, profileMetaMap);
    }

    /**
     * Accessor for the command definition document
     * @readonly
     * @private
     * @type {ICommandDefinition}
     * @memberof CommandProfileLoader
     */
    private get definition(): ICommandDefinition {
        return this.mCommandDefinition;
    }

    /**
     * Accessor for the logging object
     * @readonly
     * @private
     * @type {Logger}
     * @memberof CommandProfileLoader
     */
    private get log(): Logger {
        return this.mLog;
    }
}
