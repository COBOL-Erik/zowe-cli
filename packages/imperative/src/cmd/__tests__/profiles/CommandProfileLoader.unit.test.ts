/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

jest.mock("../../../profiles/src/BasicProfileManager");
jest.mock("../../../profiles/src/BasicProfileManagerFactory");
jest.mock("../../../utilities/src/ImperativeConfig");
jest.mock("../../../logger/src/LoggerUtils");
import { CommandProfileLoader } from "../../src/profiles/CommandProfileLoader";
import { ICommandDefinition } from "../../src/doc/ICommandDefinition";
import { BasicProfileManager } from "../../../profiles/src/BasicProfileManager";
import { TestLogger } from "../../../../__tests__/src/TestLogger";
import { CommandProfiles } from "../../src/profiles/CommandProfiles";
import { ImperativeError } from "../../../error";
import { BasicProfileManagerFactory, IProfile, IProfileLoaded } from "../../../profiles";
import { ImperativeConfig } from "../../../utilities";

const TEST_PROFILES_DIR = "/test/data/profiles/fake";

const PROFILE_BANANA_TYPE: string = "banana";
const STRAWBERRY_PROFILE_TYPE: string = "strawberry";

const SAMPLE_COMMAND_NO_PROFILE: ICommandDefinition = {
    name: PROFILE_BANANA_TYPE,
    description: "The banana command",
    type: "command"
};

const SAMPLE_COMMAND_PROFILE: ICommandDefinition = {
    name: PROFILE_BANANA_TYPE,
    description: "The banana command",
    type: "command",
    profile: {
        required: [PROFILE_BANANA_TYPE]
    }
};

const SAMPLE_COMMAND_TWO_PROFILE_TYPES: ICommandDefinition = {
    name: "bunch",
    description: "The banana command",
    type: "command",
    profile: {
        required: [PROFILE_BANANA_TYPE, STRAWBERRY_PROFILE_TYPE]
    }
};

const SAMPLE_COMMAND_TWO_PROFILE_TYPES_ONE_OPTIONAL: ICommandDefinition = {
    name: "bunch",
    description: "The banana command",
    type: "command",
    profile: {
        required: [PROFILE_BANANA_TYPE],
        optional: [STRAWBERRY_PROFILE_TYPE]
    }
};

const sampleRoot = __dirname + "/__tests__/__results__/data/";

describe("Command Profile Loader", () => {

    it("should allow us to create an instance", () => {
        const loader = CommandProfileLoader.loader({
            commandDefinition: SAMPLE_COMMAND_NO_PROFILE,
            profileManagerFactory: new BasicProfileManagerFactory(TEST_PROFILES_DIR),
            logger: TestLogger.getTestLogger()
        });
        expect(loader).toBeDefined();
    });

    it("should allow us to create an instance and load nothing", async () => {
        const loaded: CommandProfiles = await CommandProfileLoader.loader({
            commandDefinition: SAMPLE_COMMAND_NO_PROFILE,
            profileManagerFactory: new BasicProfileManagerFactory(TEST_PROFILES_DIR),
            logger: TestLogger.getTestLogger()
        }).loadProfiles({ _: undefined as any, $0: undefined as any });
        expect(loaded).toBeDefined();
    });

    it("should allow us to create an instance without a logger", () => {
        const loader = CommandProfileLoader.loader({
            commandDefinition: SAMPLE_COMMAND_NO_PROFILE,
            profileManagerFactory: new BasicProfileManagerFactory(TEST_PROFILES_DIR)
        });
        expect(loader).toBeDefined();
    });

    it("should allow us to create an instance (directly with constructor)", () => {
        const loader = new CommandProfileLoader(SAMPLE_COMMAND_NO_PROFILE,
            new BasicProfileManagerFactory(TEST_PROFILES_DIR));
        expect(loader).toBeDefined();
    });

    it("should detect a bad logger instance", () => {
        let error;
        try {
            const manager = new BasicProfileManagerFactory(TEST_PROFILES_DIR);
            let logger: any = TestLogger.getTestLogger();
            logger = {bad: "logger"};
            CommandProfileLoader.loader({
                commandDefinition: SAMPLE_COMMAND_NO_PROFILE,
                profileManagerFactory: manager,
                logger
            });
        } catch (e) {
            error = e;
        }
        expect(error).toBeDefined();
        expect(error instanceof ImperativeError).toBe(true);
        expect(error.message).toContain('Expect Error: Could not construct the profile loader. The "logger" supplied is not of type Logger.');
    });

    it("should detect missing command definitions when creating the loader", () => {
        let error;
        try {
            const manager = new BasicProfileManagerFactory(TEST_PROFILES_DIR);
            CommandProfileLoader.loader({
                commandDefinition: undefined as any,
                profileManagerFactory: manager,
                logger: TestLogger.getTestLogger()
            });
        } catch (e) {
            error = e;
        }
        expect(error).toBeDefined();
        expect(error instanceof ImperativeError).toBe(true);
        expect(error.message).toContain("Expect Error: Could not construct the profile loader. No command definition supplied.");
    });

    it("should detect missing profile manager when creating the loader", () => {
        let error;
        try {
            const manager = new BasicProfileManagerFactory(TEST_PROFILES_DIR);
            CommandProfileLoader.loader({
                commandDefinition: SAMPLE_COMMAND_NO_PROFILE,
                profileManagerFactory: undefined as any,
                logger: TestLogger.getTestLogger()
            });
        } catch (e) {
            error = e;
        }
        expect(error).toBeDefined();
        expect(error instanceof ImperativeError).toBe(true);
        expect(error.message).toContain("Expect Error: Could not construct the profile loader. No profile factory supplied.");
    });

    it("should not load old profiles when in team-config mode", async () => {
        const manager = new BasicProfileManagerFactory(TEST_PROFILES_DIR);
        const profManager = new BasicProfileManager({
            logger: TestLogger.getTestLogger(),
            profileRootDirectory: sampleRoot,
            type: PROFILE_BANANA_TYPE
        });
        manager.getManager = jest.fn((type) => {
            return profManager;
        });
        profManager.load = jest.fn((parms) => Promise.resolve({
            message: "Profile Loaded",
            type: PROFILE_BANANA_TYPE,
            name: "tasty",
            failNotFound: true,
            profile: {
                name: "tasty",
                type: PROFILE_BANANA_TYPE,
            }
        }));

        // pretend that we have a team config
        ImperativeConfig.instance.config = {
            exists: true
        } as any;

        const emptyProfileMap: Map<string, IProfile[]> = new Map<string, IProfile[]>();
        const emptyProfileMetaMap: Map<string, IProfileLoaded[]> = new Map<string, IProfileLoaded[]>();
        const noProfilesLoaded = new CommandProfiles(emptyProfileMap, emptyProfileMetaMap);

        // because we have a team config, we should load no old-scemptyProfileMaphool profiles
        const loadedCmdProfiles: CommandProfiles = await CommandProfileLoader.loader({
            commandDefinition: SAMPLE_COMMAND_PROFILE,
            profileManagerFactory: manager,
            logger: TestLogger.getTestLogger()
        }).loadProfiles({ _: undefined as any, $0: undefined as any });

        expect(loadedCmdProfiles).toEqual(noProfilesLoaded);

        // restore to not having a team config for future tests
        ImperativeConfig.instance.config = {
            exists: false
        } as any;
    });

    it("should allow us to load a required profile", async () => {
        const manager = new BasicProfileManagerFactory(TEST_PROFILES_DIR);
        const profManager = new BasicProfileManager({
            logger: TestLogger.getTestLogger(),
            profileRootDirectory: sampleRoot,
            type: PROFILE_BANANA_TYPE
        });
        manager.getManager = jest.fn((type) => {
            return profManager;
        });
        profManager.load = jest.fn((parms) => Promise.resolve({
            message: "Profile Loaded",
            type: PROFILE_BANANA_TYPE,
            name: "tasty",
            failNotFound: true,
            profile: {
                name: "tasty",
                type: PROFILE_BANANA_TYPE,
            }
        }));

        const expectedMap = new Map<string, IProfile[]>();
        expectedMap.set("banana", [{ "name": "tasty", "type": "banana" }]);

        const expectedMetaMap = new Map<string, IProfile[]>();
        expectedMetaMap.set("banana", [{
            "failNotFound": true,
            "message": "Profile Loaded",
            "name": "tasty",
            "profile": { "name": "tasty", "type": "banana" },
            "type": "banana" }
        ]);

        const response = await CommandProfileLoader.loader({
            commandDefinition: SAMPLE_COMMAND_PROFILE,
            profileManagerFactory: manager,
            logger: TestLogger.getTestLogger()
        }).loadProfiles({ _: undefined as any, $0: undefined as any });

        expect(response).toHaveProperty("mMap", expectedMap);
        expect(response).toHaveProperty("mMetaMap", expectedMetaMap);
    });

    it("should percolate the load error to the caller", async () => {
        const manager = new BasicProfileManagerFactory(TEST_PROFILES_DIR);
        const profManager = new BasicProfileManager({
            logger: TestLogger.getTestLogger(),
            profileRootDirectory: sampleRoot,
            type: PROFILE_BANANA_TYPE
        });
        manager.getManager = jest.fn((type) => {
            return profManager;
        });
        profManager.load = jest.fn((parms) => {
            throw new ImperativeError({msg: `An error occurred during the load.`});
        });
        let error;
        try {
            const response = await CommandProfileLoader.loader({
                commandDefinition: SAMPLE_COMMAND_PROFILE,
                profileManagerFactory: manager,
                logger: TestLogger.getTestLogger()
            })
                .loadProfiles({ _: undefined as any, $0: undefined as any });
        } catch (e) {
            error = e;
        }
        expect(error).toBeDefined();
        expect(error instanceof ImperativeError).toBe(true);
        expect(error.message).toMatchSnapshot();
    });

    it("should react properly if the profile manager does not return an expected result for default", async () => {
        const manager = new BasicProfileManagerFactory(TEST_PROFILES_DIR);
        const profManager = new BasicProfileManager({
            logger: TestLogger.getTestLogger(),
            profileRootDirectory: sampleRoot,
            type: PROFILE_BANANA_TYPE
        });
        manager.getManager = jest.fn((type) => {
            return profManager;
        });
        profManager.load = jest.fn((parms) => Promise.resolve({} as any ));
        let error;
        try {
            const response = await CommandProfileLoader.loader({
                commandDefinition: SAMPLE_COMMAND_PROFILE,
                profileManagerFactory: manager,
                logger: TestLogger.getTestLogger()
            })
                .loadProfiles({ _: undefined as any, $0: undefined as any });
        } catch (e) {
            error = e;
        }
        expect(error).toBeDefined();
        expect(error instanceof ImperativeError).toBe(true);
        expect(error.message).toMatchSnapshot();
    });

    it("should react properly if the profile manager does not return an expected result", async () => {
        const manager = new BasicProfileManagerFactory(TEST_PROFILES_DIR);
        const profManager = new BasicProfileManager({
            logger: TestLogger.getTestLogger(),
            profileRootDirectory: sampleRoot,
            type: PROFILE_BANANA_TYPE
        });
        manager.getManager = jest.fn((type) => {
            return profManager;
        });
        profManager.load = jest.fn((parms) => Promise.resolve({} as any));
        let error;
        try {
            const response = await CommandProfileLoader.loader({
                commandDefinition: SAMPLE_COMMAND_PROFILE,
                profileManagerFactory: manager,
                logger: TestLogger.getTestLogger()
            })
                .loadProfiles({ "_": undefined as any, "$0": undefined as any, "banana-profile": "tasty"});
        } catch (e) {
            error = e;
        }
        expect(error).toBeDefined();
        expect(error instanceof ImperativeError).toBe(true);
        expect(error.message).toMatchSnapshot();
    });

    it("should allow us to load a required profile by name", async () => {
        const manager = new BasicProfileManagerFactory(TEST_PROFILES_DIR);
        const profManager = new BasicProfileManager({
            logger: TestLogger.getTestLogger(),
            profileRootDirectory: sampleRoot,
            type: PROFILE_BANANA_TYPE
        });
        manager.getManager = jest.fn((type) => {
            return profManager;
        });
        profManager.load = jest.fn((parms) => Promise.resolve({
            message: "Profile Loaded",
            type: PROFILE_BANANA_TYPE,
            name: "tasty",
            failNotFound: true,
            profile: {
                name: "tasty",
                type: PROFILE_BANANA_TYPE,
            }
        }));

        const expectedMap = new Map<string, IProfile[]>();
        expectedMap.set("banana", [{ "name": "tasty", "type": "banana" }]);

        const expectedMetaMap = new Map<string, IProfile[]>();
        expectedMetaMap.set("banana", [{
            "failNotFound": true,
            "message": "Profile Loaded",
            "name": "tasty",
            "profile": { "name": "tasty", "type": "banana" },
            "type": "banana"
        }
        ]);

        const response = await CommandProfileLoader.loader({
            commandDefinition: SAMPLE_COMMAND_PROFILE,
            profileManagerFactory: manager,
            logger: TestLogger.getTestLogger()
        }).loadProfiles({"_": undefined as any, "$0": undefined as any, "banana-profile": "tasty"});

        expect(response).toHaveProperty("mMap", expectedMap);
        expect(response).toHaveProperty("mMetaMap", expectedMetaMap);
    });

    it("should allow us to load a required profile by name with a dependency", async () => {
        const manager = new BasicProfileManagerFactory(TEST_PROFILES_DIR);
        const profManager = new BasicProfileManager({
            logger: TestLogger.getTestLogger(),
            profileRootDirectory: sampleRoot,
            type: PROFILE_BANANA_TYPE
        });
        manager.getManager = jest.fn((type) => {
            return profManager;
        });
        profManager.load = jest.fn((parms) => Promise.resolve({
            message: "Profile Loaded",
            type: PROFILE_BANANA_TYPE,
            name: "tasty",
            failNotFound: true,
            profile: {
                name: "tasty",
                type: PROFILE_BANANA_TYPE,
            },
            dependenciesLoaded: true,
            dependencyLoadResponses: [
                {
                    message: "Profile Loaded",
                    type: STRAWBERRY_PROFILE_TYPE,
                    name: "tasty",
                    failNotFound: true,
                    profile: {
                        name: "red",
                        type: STRAWBERRY_PROFILE_TYPE,
                    },
                }
            ]
        }));

        const expectedMap = new Map<string, IProfile[]>();
        expectedMap.set("banana", [{ "name": "tasty", "type": "banana" }]);
        expectedMap.set("strawberry", [{ "name": "red", "type": "strawberry" }]);

        const expectedMetaMap = new Map<string, IProfile[]>();
        expectedMetaMap.set("banana", [{
            "dependenciesLoaded": true,
            "dependencyLoadResponses": [{
                "failNotFound": true,
                "message": "Profile Loaded",
                "name": "tasty",
                "profile": { "name": "red", "type": "strawberry" },
                "type": "strawberry"
            }],
            "failNotFound": true,
            "message": "Profile Loaded",
            "name": "tasty",
            "profile": { "name": "tasty", "type": "banana" },
            "type": "banana"
        }]);
        expectedMetaMap.set("strawberry", [{
            "failNotFound": true,
            "message": "Profile Loaded",
            "name": "tasty",
            "profile": { "name": "red", "type": "strawberry" },
            "type": "strawberry"
        }]);

        const response = await CommandProfileLoader.loader({
            commandDefinition: SAMPLE_COMMAND_PROFILE,
            profileManagerFactory: manager,
            logger: TestLogger.getTestLogger()
        }).loadProfiles({"_": undefined as any, "$0": undefined as any, "banana-profile": "tasty"});

        expect(response).toHaveProperty("mMap", expectedMap);
        expect(response).toHaveProperty("mMetaMap", expectedMetaMap);
    });

    it("should allow us to load two different required types", async () => {
        const manager = new BasicProfileManagerFactory(TEST_PROFILES_DIR);
        const profManagerBanana = new BasicProfileManager({
            logger: TestLogger.getTestLogger(),
            profileRootDirectory: sampleRoot,
            type: PROFILE_BANANA_TYPE
        });
        const profManagerStrawberry = new BasicProfileManager({
            logger: TestLogger.getTestLogger(),
            profileRootDirectory: sampleRoot,
            type: PROFILE_BANANA_TYPE
        });
        manager.getManager = jest.fn((type) => {
            if (type === PROFILE_BANANA_TYPE) {
                return profManagerBanana;
            }
            if (type === STRAWBERRY_PROFILE_TYPE) {
                return profManagerStrawberry;
            }
            return undefined as any;
        });
        profManagerBanana.load = jest.fn((parms) => Promise.resolve({
            message: "Profile Loaded",
            type: PROFILE_BANANA_TYPE,
            name: "tasty",
            failNotFound: true,
            profile: {
                name: "tasty",
                type: PROFILE_BANANA_TYPE,
            }
        }));
        profManagerStrawberry.load = jest.fn((parms) => Promise.resolve({
            message: "Profile Loaded",
            type: STRAWBERRY_PROFILE_TYPE,
            name: "tasty",
            failNotFound: true,
            profile: {
                name: "red",
                type: STRAWBERRY_PROFILE_TYPE,
            }
        }));

        const expectedMap = new Map<string, IProfile[]>();
        expectedMap.set("banana", [{ "name": "tasty", "type": "banana" }]);
        expectedMap.set("strawberry", [{ "name": "red", "type": "strawberry" }]);

        const expectedMetaMap = new Map<string, IProfile[]>();
        expectedMetaMap.set("banana", [{
            "failNotFound": true,
            "message": "Profile Loaded",
            "name": "tasty",
            "profile": { "name": "tasty", "type": "banana" },
            "type": "banana"
        }]);
        expectedMetaMap.set("strawberry", [{
            "failNotFound": true,
            "message": "Profile Loaded",
            "name": "tasty",
            "profile": { "name": "red", "type": "strawberry" },
            "type": "strawberry"
        }]);

        const response = await CommandProfileLoader.loader({
            commandDefinition: SAMPLE_COMMAND_TWO_PROFILE_TYPES,
            profileManagerFactory: manager,
            logger: TestLogger.getTestLogger()
        }).loadProfiles({"_": undefined as any, "$0": undefined as any, "banana-profile": "tasty", "strawberry-profile": "red"});

        expect(response).toHaveProperty("mMap", expectedMap);
        expect(response).toHaveProperty("mMetaMap", expectedMetaMap);
    });

    it("should percolate the error if a required profile for one type is not found", async () => {
        const manager = new BasicProfileManagerFactory(TEST_PROFILES_DIR);
        const profManagerBanana = new BasicProfileManager({
            logger: TestLogger.getTestLogger(),
            profileRootDirectory: sampleRoot,
            type: PROFILE_BANANA_TYPE
        });
        const profManagerStrawberry = new BasicProfileManager({
            logger: TestLogger.getTestLogger(),
            profileRootDirectory: sampleRoot,
            type: PROFILE_BANANA_TYPE
        });
        manager.getManager = jest.fn((type) => {
            if (type === PROFILE_BANANA_TYPE) {
                return profManagerBanana;
            }
            if (type === STRAWBERRY_PROFILE_TYPE) {
                return profManagerStrawberry;
            }
            return undefined as any;
        });
        profManagerBanana.load = jest.fn((parms) => Promise.resolve({
            message: "Profile Loaded",
            type: PROFILE_BANANA_TYPE,
            name: "tasty",
            failNotFound: true,
            profile: {
                name: "tasty",
                type: PROFILE_BANANA_TYPE,
            }
        }));
        profManagerStrawberry.load = jest.fn((parms) => {
            throw new ImperativeError({msg: `Not found`});
        });
        let error;
        try {
            const response = await CommandProfileLoader.loader({
                commandDefinition: SAMPLE_COMMAND_TWO_PROFILE_TYPES,
                profileManagerFactory: manager,
                logger: TestLogger.getTestLogger()
            })
                .loadProfiles({"_": undefined as any, "$0": undefined as any, "banana-profile": "tasty", "strawberry-profile": "red"});
        } catch (e) {
            error = e;
        }
        expect(error).toBeDefined();
        expect(error instanceof ImperativeError).toBe(true);
        expect(error.message).toMatchSnapshot();
    });

    it("should handle multiple loads of the same type", async () => {
        const manager = new BasicProfileManagerFactory(TEST_PROFILES_DIR);
        const profManagerBanana = new BasicProfileManager({
            logger: TestLogger.getTestLogger(),
            profileRootDirectory: sampleRoot,
            type: PROFILE_BANANA_TYPE
        });
        manager.getManager = jest.fn((type) => {
            if (type === PROFILE_BANANA_TYPE) {
                return profManagerBanana;
            }
            return undefined as any;
        });
        profManagerBanana.load = jest.fn((parms) => Promise.resolve({
            message: "Profile Loaded",
            type: PROFILE_BANANA_TYPE,
            name: "tasty",
            failNotFound: true,
            profile: {
                name: "tasty",
                type: PROFILE_BANANA_TYPE,
            },
            dependenciesLoaded: true,
            dependencyLoadResponses: [
                {
                    message: "Profile Loaded",
                    type: PROFILE_BANANA_TYPE,
                    name: "great",
                    failNotFound: true,
                    profile: {
                        name: "great",
                        type: PROFILE_BANANA_TYPE,
                    },
                    dependenciesLoaded: true,
                    dependencyLoadResponses: [
                        {
                            message: "Profile Loaded",
                            type: PROFILE_BANANA_TYPE,
                            name: "awesome",
                            failNotFound: true,
                            profile: {
                                name: "awesome",
                                type: PROFILE_BANANA_TYPE,
                            },
                        }
                    ]
                }
            ]
        }));

        const expectedMap = new Map<string, IProfile[]>();
        expectedMap.set("banana", [
            { "name": "tasty", "type": "banana" },
            { "name": "great", "type": "banana" },
            { "name": "awesome", "type": "banana" }
        ]);

        const expectedMetaMap = new Map<string, IProfile[]>();
        expectedMetaMap.set("banana", [
            {
                "dependenciesLoaded": true,
                "dependencyLoadResponses": [{
                    "dependenciesLoaded": true,
                    "failNotFound": true,
                    "message": "Profile Loaded",
                    "name": "great",
                    "profile": { "name": "great", "type": "banana" },
                    "type": "banana"
                }],
                "failNotFound": true,
                "message": "Profile Loaded",
                "name": "tasty",
                "profile": { "name": "tasty", "type": "banana" },
                "type": "banana"
            },
            {
                "dependenciesLoaded": true,
                "failNotFound": true,
                "message": "Profile Loaded",
                "name": "great",
                "profile": { "name": "great", "type": "banana" },
                "type": "banana"
            },
            {
                "failNotFound": true,
                "message": "Profile Loaded",
                "name": "awesome",
                "profile": { "name": "awesome", "type": "banana" },
                "type": "banana"
            }
        ]);

        // commandDefinition: SAMPLE_COMMAND_TWO_PROFILE_TYPES_ONE_OPTIONAL,
        const response = await CommandProfileLoader.loader({
            commandDefinition: SAMPLE_COMMAND_PROFILE,
            profileManagerFactory: manager,
            logger: TestLogger.getTestLogger()
        }).loadProfiles({"_": undefined as any, "$0": undefined as any, "banana-profile": "tasty"});

        expect(response).toHaveProperty("mMap", expectedMap);
        expect(response).toHaveProperty("mMetaMap", expectedMetaMap);
    });

    it("should handle load of required and optional profiles", async () => {
        const manager = new BasicProfileManagerFactory(TEST_PROFILES_DIR);
        const profManagerBanana = new BasicProfileManager({
            logger: TestLogger.getTestLogger(),
            profileRootDirectory: sampleRoot,
            type: PROFILE_BANANA_TYPE
        });
        const profManagerStrawberry = new BasicProfileManager({
            logger: TestLogger.getTestLogger(),
            profileRootDirectory: sampleRoot,
            type: STRAWBERRY_PROFILE_TYPE
        });
        manager.getManager = jest.fn((type) => {
            if (type === PROFILE_BANANA_TYPE) {
                return profManagerBanana;
            }
            if (type === STRAWBERRY_PROFILE_TYPE) {
                return profManagerStrawberry;
            }
            return undefined as any;
        });
        profManagerBanana.load = jest.fn((parms) => Promise.resolve({
            message: "Profile Loaded",
            type: PROFILE_BANANA_TYPE,
            name: "tasty",
            failNotFound: true,
            profile: {
                name: "tasty",
                type: PROFILE_BANANA_TYPE,
            }
        }));
        profManagerStrawberry.load = jest.fn((parms) => Promise.resolve({
            message: "Profile Loaded",
            type: STRAWBERRY_PROFILE_TYPE,
            name: "tasty",
            failNotFound: true,
            profile: {
                name: "red",
                type: STRAWBERRY_PROFILE_TYPE,
            }
        }));

        const expectedMap = new Map<string, IProfile[]>();
        expectedMap.set("banana", [
            { "name": "tasty", "type": "banana" }
        ]);
        expectedMap.set("strawberry", [
            { "name": "red", "type": "strawberry" }
        ]);

        const expectedMetaMap = new Map<string, IProfile[]>();
        expectedMetaMap.set("banana", [
            {
                "failNotFound": true,
                "message": "Profile Loaded",
                "name": "tasty",
                "profile": { "name": "tasty", "type": "banana" },
                "type": "banana"
            }
        ]);
        expectedMetaMap.set("strawberry", [
            {
                "failNotFound": true,
                "message": "Profile Loaded",
                "name": "tasty",
                "profile": { "name": "red", "type": "strawberry" },
                "type": "strawberry"
            }
        ]);

        const response = await CommandProfileLoader.loader({
            commandDefinition: SAMPLE_COMMAND_TWO_PROFILE_TYPES_ONE_OPTIONAL,
            profileManagerFactory: manager,
            logger: TestLogger.getTestLogger()
        }).loadProfiles({"_": undefined as any, "$0": undefined as any, "banana-profile": "tasty", "strawberry-profile": "red"});

        expect(response).toHaveProperty("mMap", expectedMap);
        expect(response).toHaveProperty("mMetaMap", expectedMetaMap);
    });
});
