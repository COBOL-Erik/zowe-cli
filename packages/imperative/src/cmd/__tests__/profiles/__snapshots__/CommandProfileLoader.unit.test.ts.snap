// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Command Profile Loader should percolate the error if a required profile for one type is not found 1`] = `"Not found"`;

exports[`Command Profile Loader should percolate the load error to the caller 1`] = `"An error occurred during the load."`;

exports[`Command Profile Loader should react properly if the profile manager does not return an expected result 1`] = `"Unexpected internal load error: The profile \\"tasty\\" was not loaded by the profile manager."`;

exports[`Command Profile Loader should react properly if the profile manager does not return an expected result for default 1`] = `"Unexpected internal load error: The profile \\"default requested\\" was not loaded by the profile manager."`;
